File: painradar/signals.py
"""Transparent, rule-based pain signal detection.

Detection is purely phrase/pattern based (no ML, no external calls) so that every
match can be explained back to the user with the exact phrase that triggered it.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Dict, List, Set

from .textnorm import normalize_text

# Each category maps to a list of regex patterns (already lowercase, matched against
# normalized text). Patterns are intentionally simple and readable so the rules stay
# auditable by a human.
PAIN_CATEGORIES: Dict[str, List[str]] = {
    "manual_work": [
        r"\bmanually\b",
        r"\bmanual work\b",
        r"\bmanual process\b",
        r"\bby hand\b",
        r"\bcopy[- ]?paste\b",
        r"\bevery (week|day|single day)\b.*\b(manual|copy|type|enter)\w*\b",
        r"\brepetitive\b",
    ],
    "expensive_tools": [
        r"\btoo expensive\b",
        r"\bexpensive\b",
        r"\bcost(s|ing)? too much\b",
        r"\b(way )?overpriced\b",
        r"\benterprise[- ]priced\b",
        r"\$\d+\s*/?\s*(month|mo|year)\b",
    ],
    "missing_solution": [
        r"\bno (good )?solution\b",
        r"\bno affordable\b",
        r"\bmissing solution\b",
        r"\bdoesn'?t exist\b",
        r"\bnothing (out there|exists)\b",
        r"\bcan'?t find (a|any) (tool|solution)\b",
    ],
    "frustration": [
        r"\bfrustrat\w*\b",
        r"\bexhaust\w*\b",
        r"\bdraining\b",
        r"\bso (tired|sick) of\b",
        r"\bkilling me\b",
        r"\bdestroying my\b",
    ],
    "willingness_to_pay": [
        r"\bwould pay\b",
        r"\bwilling to pay\b",
        r"\bi'?d pay\b",
        r"\bhappily pay\b",
        r"\bpay for (a|any|something)\b",
        r"\btake my money\b",
    ],
}

_COMPILED: Dict[str, List[re.Pattern]] = {
    category: [re.compile(p) for p in patterns]
    for category, patterns in PAIN_CATEGORIES.items()
}


@dataclass
class PainSignalResult:
    matched_categories: Set[str] = field(default_factory=set)
    matched_phrases: Dict[str, List[str]] = field(default_factory=dict)

    @property
    def diversity_score(self) -> int:
        return len(self.matched_categories)


def detect_pain_signals(text: str) -> PainSignalResult:
    normalized = normalize_text(text)
    result = PainSignalResult()
    for category, patterns in _COMPILED.items():
        matches = [p.pattern for p in patterns if p.search(normalized)]
        if matches:
            result.matched_categories.add(category)
            result.matched_phrases[category] = matches
    return result

File: src/signals.ts
import { normalizeText } from "./textnorm.js";

export const PAIN_CATEGORIES: Readonly<Record<string, readonly string[]>> = {
  manual_work: [
    "\\bmanually\\b",
    "\\bmanual work\\b",
    "\\bmanual process\\b",
    "\\bby hand\\b",
    "\\bcopy[- ]?paste\\b",
    "\\bevery (week|day|single day)\\b.*\\b(manual|copy|type|enter)\\w*\\b",
    "\\brepetitive\\b",
  ],
  expensive_tools: [
    "\\btoo expensive\\b",
    "\\bexpensive\\b",
    "\\bcost(s|ing)? too much\\b",
    "\\b(way )?overpriced\\b",
    "\\benterprise[- ]priced\\b",
    "\\$\\d+\\s*/?\\s*(month|mo|year)\\b",
  ],
  missing_solution: [
    "\\bno (good )?solution\\b",
    "\\bno affordable\\b",
    "\\bmissing solution\\b",
    "\\bdoesn'?t exist\\b",
    "\\bnothing (out there|exists)\\b",
    "\\bcan'?t find (a|any) (tool|solution)\\b",
  ],
  frustration: [
    "\\bfrustrat\\w*\\b",
    "\\bexhaust\\w*\\b",
    "\\bdraining\\b",
    "\\bso (tired|sick) of\\b",
    "\\bkilling me\\b",
    "\\bdestroying my\\b",
  ],
  willingness_to_pay: [
    "\\bwould pay\\b",
    "\\bwilling to pay\\b",
    "\\bi'?d pay\\b",
    "\\bhappily pay\\b",
    "\\bpay for (a|any|something)\\b",
    "\\btake my money\\b",
  ],
};

const COMPILED: Record<string, RegExp[]> = Object.fromEntries(
  Object.entries(PAIN_CATEGORIES).map(([category, patterns]) => [
    category,
    patterns.map((p) => new RegExp(p)),
  ]),
);

export class PainSignalResult {
  matchedCategories: Set<string> = new Set();
  matchedPhrases: Record<string, string[]> = {};

  get diversityScore(): number {
    return this.matchedCategories.size;
  }
}

export function detectPainSignals(text: string): PainSignalResult {
  const normalized = normalizeText(text);
  const result = new PainSignalResult();
  for (const [category, patterns] of Object.entries(COMPILED)) {
    const matches = patterns.filter((p) => p.test(normalized)).map((p) => p.source);
    if (matches.length > 0) {
      result.matchedCategories.add(category);
      result.matchedPhrases[category] = matches;
    }
  }
  return result;
}
